src/super_gradients/training/sg_trainer/sg_trainer.py

```diff
@@ -483,7 +483,31 @@ def _train_epoch(self, context: PhaseContext, silent_mode: bool = False) -> tupl
 
             context.update_context(loss_avg_meter=loss_avg_meter, metrics_compute_fn=self.train_metrics)
 
-            for batch_idx, batch_items in enumerate(progress_bar_train_loader):
+            class PrefetchIterator:
+                def __init__(self, iterator):
+                    self.iterator = iterator
+                    import concurrent.futures
+                    self.executor = concurrent.futures.ThreadPoolExecutor(max_workers=1)
+                    self.prefetch()
+
+                def prefetch(self):
+                    self.prefetch_future = self.executor.submit(self._prefetch)
+
+                def _prefetch(self):
+                    return next(self.iterator)
+
+                def __iter__(self):
+                    return self
+
+                def __next__(self):
+                    value = self.prefetch_future.result()
+                    self.prefetch()
+                    return value
+
+                def close(self):
```

Comment:
Who is calling `close` method? Maybe we should have some __enter__ / __exit__ methods to ensure we do proper cleanup of the ThreadPoolExecutor.

Reply:
Close is automatically called when the for loop using the iterator breaks.

Reply:
Actually I was mistaken, the python language construct to release an iterator is `__del__`, not `close`.

Reply:
Changed how the iterator is created/managed so this code is defunct.